ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -488,40 +494,76 @@ public void close() {
+              .isTtlUpdate()) {
+            // This DELETE is not valid, when the final state of this storeKey is
+            // 1. UNDELETE, or
+            // 2. DELETE, but the current lifeVersion is not the same, or
+            // 3. TTL_UPDATE
+            continue;
+          }
+        } else {
+          long operationTimeInMs =
+              indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
+                  : indexValue.getOperationTimeInMs();
+          keyFinalStates.put(indexEntry.getKey(),
+              new IndexFinalState(indexValue.getFlags(), operationTimeInMs, indexValue.getLifeVersion()));
+        }
         validIndexEntryAction.accept(indexEntry);
-        if (!isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs)) {
+      } else if (indexValue.isUndelete()) {
+        if (keyFinalStates.containsKey(key)) {
+          IndexFinalState state = keyFinalStates.get(key);
+          if (state.isTtlUpdate()) {
+            indexValue.setExpiresAtMs(Utils.Infinite_Time);
+          }
+          if (state.isDelete() || (state.getLifeVersion() != indexValue.getLifeVersion())) {
+            // This UNDELETE is not valid, when the final state of this storeKey is
+            // 1. DELETE, or
+            // 2. the current lifeVersion is not the same
+            continue;
+          }
+        } else {
           long operationTimeInMs =
               indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
                   : indexValue.getOperationTimeInMs();
-          deletedKeys.put(indexEntry.getKey(), operationTimeInMs);
+          keyFinalStates.put(indexEntry.getKey(),
+              new IndexFinalState(indexValue.getFlags(), operationTimeInMs, indexValue.getLifeVersion()));
+        }
+        if (!isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs)) {
+          validIndexEntryAction.accept(indexEntry);
         }
       } else if (indexValue.isTtlUpdate()) {
-        if (isTtlUpdateEntryValid(indexEntry.getKey(), indexValue, referenceTimeInMs, deletedKeys)) {
+        if (isTtlUpdateEntryValid(key, indexValue, referenceTimeInMs, keyFinalStates)) {
           validIndexEntryAction.accept(indexEntry);
         }
       } else {
-        // This is a put record
-        if (isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs)) {
-          // try to update the expiration time for this put if there is an ttl update
-          IndexValue newValue = index.findKey(indexEntry.getKey());
-          if (newValue != null) {
-            indexValue.setExpiresAtMs(newValue.getExpiresAtMs());
-          }
+        IndexFinalState finalState = removeFromStates ? keyFinalStates.remove(key) : keyFinalStates.get(key);
+        if (finalState != null && finalState.isDelete() && finalState.getOperationTime() < referenceTimeInMs) {
+          // Put is deleted before reference time, it's not valid.
```

Comment:
I don't understand why we need check `inalState.getOperationTime() < referenceTimeInMs`.  I think once final state is delete, this put is no longer valid.

Reply:
it's not just that. when the final state is delete, but it's still within the retention duration, say it's deleted no more than 7 days ago, then the put is still a valid record. It's only after 7 days, the put would be invalid.